index.js

```diff
@@ -19,6 +20,14 @@ module.exports = (fn, opts) => {
 		cachePromiseRejection: false
 	}, opts);
 
+	const fnKey = JSON.stringify(opts);
```

Comment:
`JSON.stringify` cannot handle `new Map()`, so if `options.cache` differs between calls, it will still use the same instance.

```js
JSON.stringify({foo: new Map([['a', 'b']])})
//=> '{"foo":{}}'
```

Reply:
Agreed. Ignoring the `cache` option is _kind of intentional_ though - the intent is to keep `mem(fn)` idempotent; giving `mem(fn)` its own cache each time is redundant. It could even be said giving `mem(fn)` and `mem(fn, { cachePromiseRejection: false })` their own cache is also redundant.

There's also another issue we have; `options = Object.assign(...)` will always generate a new reference, and thus just putting `options` in a Map will mean `mem` will never memoize. So we need some kind of mechanism to cache these options, but yes - as you say - JSON.stringify is insufficient.

One way around this could be to store each option in a Map - this is a little clunky, but it also gives us the benefit of ignoring the `cache` as part of the memoizing.

```js
	if (fnStore.has(fn)) {
		// Check if a previously memoized `fn` was given the same cacheKey option:
		const cacheKeyStore = fnStore.get(fn);
		if (cacheKeyStore.has(options.cacheKey)) {
			// Check if a previously memoized `fn` was given the same cachePromiseRejection option:
			const cachePromiseRejectionStore = cacheKeyStore.get(options.cacheKey)
			if (cachePromiseRejectionStore.has(options.cachePromiseRejection)) {
				// We have a memoized version stored with all the same options, return it
				return cachePromiseRejectionStore.get(options.cachePromiseRejection)
			}
		}
	}
```